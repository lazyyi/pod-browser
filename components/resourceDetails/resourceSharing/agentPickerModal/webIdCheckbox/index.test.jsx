/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import userEvent from "@testing-library/user-event";
import { renderWithTheme } from "../../../../../__testUtils/withTheme";
import useContactProfile from "../../../../../src/hooks/useContactProfile";
import WebIdCheckbox, { TESTCAFE_ID_WEBID_CHECKBOX } from "./index";

jest.mock("../../../../../src/hooks/useContactProfile");

const webId = "https://somewebid.com";

describe("WebIdCheckbox", () => {
  const value = webId;
  const index = 1;
  const addingWebId = false;
  const toggleCheckbox = jest.fn();
  const newAgentsWebIds = [];
  const webIdsInPermissions = [];
  const webIdsToDelete = [];
  test("renders a checkbox with the corrext value", () => {
    useContactProfile.mockReturnValue({ data: null });

    const { asFragment, getByTestId } = renderWithTheme(
      <WebIdCheckbox
        value={value}
        index={index}
        addingWebId={addingWebId}
        toggleCheckbox={toggleCheckbox}
        newAgentsWebIds={newAgentsWebIds}
        webIdsInPermissions={webIdsInPermissions}
        webIdsToDelete={webIdsToDelete}
      />
    );
    expect(asFragment()).toMatchSnapshot();
    const checkbox = getByTestId(TESTCAFE_ID_WEBID_CHECKBOX);
    expect(checkbox).toBeDefined();
    expect(checkbox).toHaveAttribute("value", webId);
  });
  test("checkbox has the correct value if profile is available", () => {
    useContactProfile.mockReturnValue({
      data: { webId: "https://example.org" },
    });
    const { getByTestId } = renderWithTheme(
      <WebIdCheckbox
        value={null}
        index={index}
        addingWebId={addingWebId}
        toggleCheckbox={toggleCheckbox}
        newAgentsWebIds={newAgentsWebIds}
        webIdsInPermissions={webIdsInPermissions}
        webIdsToDelete={webIdsToDelete}
      />
    );
    const checkbox = getByTestId(TESTCAFE_ID_WEBID_CHECKBOX);
    expect(checkbox).toBeDefined();
    expect(checkbox).toHaveAttribute("value", "https://example.org");
  });
  test("checkbox has a null value if value is unavailable and it's the first row", () => {
    useContactProfile.mockReturnValue({
      data: null,
    });
    const { getByTestId } = renderWithTheme(
      <WebIdCheckbox
        value={null}
        index={0}
        addingWebId={addingWebId}
        toggleCheckbox={toggleCheckbox}
        newAgentsWebIds={newAgentsWebIds}
        webIdsInPermissions={webIdsInPermissions}
        webIdsToDelete={webIdsToDelete}
      />
    );
    const checkbox = getByTestId(TESTCAFE_ID_WEBID_CHECKBOX);
    expect(checkbox).toBeDefined();
    expect(checkbox).toHaveAttribute("value", "");
  });
  test("checkbox is checked if agent is already in permissions", () => {
    useContactProfile.mockReturnValue({
      data: { webId },
    });
    const { getByTestId } = renderWithTheme(
      <WebIdCheckbox
        value={null}
        index={0}
        addingWebId={addingWebId}
        toggleCheckbox={toggleCheckbox}
        newAgentsWebIds={newAgentsWebIds}
        webIdsInPermissions={[webId]}
        webIdsToDelete={webIdsToDelete}
      />
    );
    const checkbox = getByTestId(TESTCAFE_ID_WEBID_CHECKBOX);
    expect(checkbox).toBeDefined();
    expect(checkbox).toBeChecked();
  });
  test("calls toggleCheckbox on click with the correct values", () => {
    useContactProfile.mockReturnValue({
      data: { webId },
    });
    const { getByTestId } = renderWithTheme(
      <WebIdCheckbox
        value={null}
        index={index}
        addingWebId={addingWebId}
        toggleCheckbox={toggleCheckbox}
        newAgentsWebIds={newAgentsWebIds}
        webIdsInPermissions={[webId]}
        webIdsToDelete={webIdsToDelete}
      />
    );
    const checkbox = getByTestId(TESTCAFE_ID_WEBID_CHECKBOX);
    userEvent.click(checkbox);
    expect(toggleCheckbox).toHaveBeenCalledWith(
      expect.any(Object),
      index,
      webId
    );
  });
});
