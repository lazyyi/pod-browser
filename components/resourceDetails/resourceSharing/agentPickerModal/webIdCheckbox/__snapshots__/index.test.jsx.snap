// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebIdCheckbox renders a checkbox with the corrext value 1`] = `
<DocumentFragment>
  <span
    aria-disabled="false"
    class="PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-root PodBrowser-colorPrimary PodBrowser-colorPrimary"
  >
    <span
      class="PodBrowser-label"
    >
      <input
        class="PodBrowser-input"
        data-indeterminate="false"
        data-testid="webid-checkbox"
        type="checkbox"
        value="https://somewebid.com"
      />
      <svg
        aria-hidden="true"
        class="PodBrowser-root"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
        />
      </svg>
    </span>
    <span
      class="PodBrowser-root"
    />
  </span>
</DocumentFragment>
`;
